testing/unit/munit-open-close.cpp

```diff
@@ -0,0 +1,158 @@
+//   and/or other materials provided with the distribution.
+// * Neither the name  of Intel Corporation  nor the names of its contributors
+//   may be used to  endorse or promote  products derived  from this  software
+//   without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
+// IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+#include <opae/access.h>
+
+#ifdef __cplusplus
+}
+#endif
+
+#include "gtest/gtest.h"
+#include "test_system.h"
+#include "types_int.h"
+
+using namespace opae::testing;
+
+class openclose_c_p
+    : public ::testing::TestWithParam<std::string> {
+ protected:
+  openclose_c_p() : tmpsysfs_("mocksys-XXXXXX"){}
+
+  virtual void SetUp() override {
+    ASSERT_TRUE(test_platform::exists(GetParam()));
+    platform_ = test_platform::get(GetParam());
+    system_ = test_system::instance();
+    system_->initialize();
+    tmpsysfs_ = system_->prepare_syfs(platform_);
+
+    ASSERT_EQ(fpgaGetProperties(nullptr, &filter_), FPGA_OK);
+    ASSERT_EQ(fpgaPropertiesSetObjectType(filter_, FPGA_DEVICE), FPGA_OK);
+    ASSERT_EQ(fpgaEnumerate(&filter_, 1, tokens_.data(), tokens_.size(),
+                            &num_matches_),
+              FPGA_OK);
+  }
+
+  virtual void TearDown() override {
+    EXPECT_EQ(fpgaDestroyProperties(&filter_), FPGA_OK);
+    if (!tmpsysfs_.empty() && tmpsysfs_.size() > 1) {
+      std::string cmd = "rm -rf " + tmpsysfs_;
+      std::system(cmd.c_str());
+    }
+    system_->finalize();
+  }
+
+  std::string tmpsysfs_;
+  fpga_properties filter_;
+  std::array<fpga_token, 2> tokens_;
+  uint32_t num_matches_;
+  test_platform platform_;
+  test_system *system_;
+};
+
+
+
+
+/**
+ * @test       open_01
+ *
+ * @brief      When the fpga_handle * parameter to fpgaOpen is NULL, the
+ *             function returns FPGA_INVALID_PARAM.
+ */
+TEST_P(openclose_c_p, open_01) {
+  fpga_token tok = NULL;
+  fpga_result res;
+  res = fpgaOpen(tok, NULL, 0);
+  ASSERT_EQ(FPGA_INVALID_PARAM, res);
+}
+
+/**
+ * @test       open_02
+ *
+ * @brief      When the fpga_handle * parameter to fpgaOpen is NULL, the
+ *             function returns FPGA_INVALID_PARAM.
+ */
+TEST_P(openclose_c_p, open_02) {
```

Comment:
open_01 and open_02 appear to be the same test.  Maybe a copy/paste glitch?  I think open_02 probably needs to verify another condition.

Reply:
good catch.

Reply:
Actually, the first test tests for the token when is initialized to NULL and the second test for an declared token.

Reply:
I see the difference now.  Actually, open_02 is invalid, because tok is uninitialized.  We cannot protect against this scenario.

Reply:
ok, I will leave this test out.

Reply:
Just a thought - you *can* test this by:

char tmp[256] = { 0 };
fpga_handle fpga = (fpga_handle) &tmp[0];

The fpgaOpen() checks the signature in the handle.  Just a thought.

Reply:
Actually, if there is no valid pthread_mutex_t in the underlying object, we'll segfault inside pthread_mutex_lock().  That's the root of why we can't guard against uninitialized handles.  There is race condition between the validity of the mutex object and the magic.  Which do you do first, lock the mutex or check the magic?  If you check the magic first, you're not thread safe.  If you lock the mutex first you risk attempting to lock a garbage mutex object..

Reply:
This is an open, and there's no thread protection on tokens.  Perhaps that is an omission in the token struct, but today, there's no mutex protection present for tokens.